modules/substrate/src/verifier.rs

```diff
@@ -0,0 +1,764 @@
+				height == *header.number()
+			} else {
+				false
+			}
+		};
+
+		let is_finalized = false;
+		self.storage.write_header(&ImportedHeader {
+			header,
+			requires_justification,
+			is_finalized,
+		});
+
+		Ok(())
+	}
+
+	/// Verify that a previously imported header can be finalized with the given Grandpa finality
+	/// proof. If the header enacts an authority set change the change will be applied once the
+	/// header has been finalized.
+	pub fn verify_finality(&mut self, hash: H::Hash, proof: FinalityProof) -> Result<(), FinalizationError> {
+		// Make sure that we've previously imported this header
+		let header = self
+			.storage
+			.header_by_hash(hash)
+			.ok_or(FinalizationError::UnknownHeader)?;
+
+		// We don't want to finalize an ancestor of an already finalized
+		// header, this would be inconsistent
+		let last_finalized = self.storage.best_finalized_header();
+		if header.number() <= last_finalized.number() {
+			return Err(FinalizationError::OldHeader);
+		}
+
+		let current_authority_set = self.storage.current_authority_set();
+		let is_finalized = prove_finality(&header, &current_authority_set, proof);
+		if !is_finalized {
+			return Err(FinalizationError::UnfinalizedHeader);
+		}
+
+		let mut finalized_headers =
+			if let Some(ancestors) = headers_between(&self.storage, last_finalized, header.clone()) {
+				// Since we only try and finalize headers with a height strictly greater
+				// than `best_finalized` if `headers_between` returns Some we must have
+				// at least one element. If we don't something's gone wrong, so best
+				// to die before we write to storage.
+				assert_eq!(ancestors.is_empty(), false);
+
+				// Check if any of our ancestors `requires_justification` a.k.a schedule authority
+				// set changes. If they're still waiting to be finalized we must reject this
+				// justification.
+				//
+				// We do this because it is important to to import justifications _in order_,
+				// otherwise we risk finalizing headers on competing chains.
+				let requires_justification = ancestors.iter().skip(1).find(|h| h.requires_justification);
+				if requires_justification.is_some() {
+					return Err(FinalizationError::PrematureJustification);
+				}
+
+				ancestors
+			} else {
+				return Err(FinalizationError::AncestryCheckFailed);
+			};
+
+		// If the current header was marked as `requires_justification` it means that it enacts a
+		// new authority set change. When we finalize the header we need to update the current
+		// authority set.
+		if header.requires_justification {
+			self.storage.enact_authority_set();
+		}
+
+		for header in finalized_headers.iter_mut() {
+			header.is_finalized = true;
+			header.requires_justification = false;
+			self.storage.write_header(header);
+		}
+
+		let best_finalized = finalized_headers
+			.first()
+			.expect("Asserted that `ancestors` (now `last_finalized`) had at least one element.");
+		self.storage.update_best_finalized((*best_finalized).hash());
```

Comment:
```suggestion
		self.storage.update_best_finalized(hash);
```

Iiuc `best_finalized` is always the `header`, right?